src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,365 @@
+        /// </summary>
+        /// /// <remarks>
+        /// The first Document Symbol in the array is considered the root node.
+        /// </remarks>
+        private static DocumentSymbol CreateDocumentSymbolTree(ImmutableArray<DocumentSymbol> documentSymbols)
+        {
+            var node = documentSymbols.First();
+            var childDocumentSymbols = documentSymbols.RemoveAt(0);
+            node.Children = GroupDocumentSymbolTrees(childDocumentSymbols)
+                .Select(group => CreateDocumentSymbolTree(group))
+                .ToArray();
+            return node;
+        }
+
+        /// <summary>
+        /// Converts an array of type DocumentSymbol to an array of type DocumentSymbolViewModel.
+        /// </summary>
+        public static ImmutableArray<DocumentSymbolViewModel> GetDocumentSymbolModels(DocumentSymbol[] documentSymbols)
+        {
+            var documentSymbolModels = ArrayBuilder<DocumentSymbolViewModel>.GetInstance();
+            if (documentSymbols.Length == 0)
+            {
+                return documentSymbolModels.ToImmutableAndFree();
+            }
+
+            foreach (var documentSymbol in documentSymbols)
+            {
+                var documentSymbolModel = new DocumentSymbolViewModel(documentSymbol);
+                if (documentSymbol.Children is not null)
+                    documentSymbolModel.Children = GetDocumentSymbolModels(documentSymbol.Children);
+                documentSymbolModels.Add(documentSymbolModel);
+            }
+
+            return documentSymbolModels.ToImmutableAndFree();
+        }
+
+        public static async Task<JToken?> DocumentSymbolsRequestAsync(
+            ITextBuffer textBuffer,
+            ILanguageServiceBrokerShim languageServiceBroker,
+            string? textViewFilePath,
+            CancellationToken cancellationToken)
+        {
+            if (textViewFilePath is null)
+                return null;
+
+            var parameterFactory = new DocumentSymbolParams()
+            {
+                TextDocument = new TextDocumentIdentifier()
+                {
+                    Uri = new Uri(textViewFilePath)
+                }
+            };
+
+            // TODO: proper workaround such that context.ClientCapabilities?.TextDocument?.DocumentSymbol?.HierarchicalDocumentSymbolSupport == true
+            return await languageServiceBroker.RequestAsync(
+                textBuffer: textBuffer,
+                method: Methods.TextDocumentDocumentSymbolName,
+                capabilitiesFilter: (JToken x) => true,
+                languageServerName: WellKnownLspServerKinds.AlwaysActiveVSLspServer.ToUserVisibleString(),
+                parameterFactory: _ => JToken.FromObject(parameterFactory),
+                cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
+
+        public static void SetIsExpanded(IEnumerable<DocumentSymbolViewModel> documentSymbolModels, bool isExpanded)
```

Comment:
I wonder (and this might allude to some of other of @CyrusNajmabadi's comments) is whether we're better off just having a class that represents the overall model of holding onto the top-level list of symbols, and then we can have better methods on that.